### ISSUE TITLE
History substring search

### ISSUE THREAD

**Issue by jmiguelv** [bhilburn/powerlevel9k#131]
When I use this theme the the history substring search plugin does not work. Any ideas why or has anyone seen this problem before?

Thank you

**Comment by bhilburn** [comment]
@jmiguelv - I actually use history-substring-search, and it works fine.

I have had history-substring-search break a number of times, and it is always the result of an update to another plugin I'm using. I've found that changing the order of the plugin loads often fixes the problem. This is what I'm using right now:

```
plugins=(zsh-syntax-highlighting gitfast git-extras svn history-substring-search colored-man colorize copyfile copydir cp dircycle command-not-found)
```

**Comment by bhilburn** [comment]
It could also be your keybindings:

```
### KEY BINDINGS ############################################################
zmodload zsh/terminfo
bindkey "$terminfo[kcuu1]" history-substring-search-up
bindkey "$terminfo[kcud1]" history-substring-search-down

# Have to do this again for tmux
bindkey '^[[B' history-substring-search-down
bindkey '^[[A' history-substring-search-up

## Vim key bindings
bindkey '^?' backward-delete-char
bindkey '^R' history-incremental-search-backward
bindkey '^S' history-incremental-search-forward
bindkey '^P' history-search-backward
bindkey '^N' history-search-forward
```

**Comment by jmiguelv** [comment]
Thank you for the suggestions @bhilburn. The plugin works with other oh-my-zsh themes though. I have also tried using just the `history-substring-search` plugin, but not luck. Here is my configuration:
```zsh
export ZSH=$HOME/.oh-my-zsh

ZSH_THEME="powerlevel9k/powerlevel9k"

export DISABLE_AUTO_TITLE="true"

plugins=(brew colorize command-not-found fabric git git-extras history-substring-search last-working-dir mercurial osx pip python sudo svn-fast-info vagrant zsh-syntax-highlighting)

export PATH=$HOME/bin:/usr/local/bin:/usr/local/sbin:$PATH

source $ZSH/oh-my-zsh.sh

DEFAULT_USER=jvieira

POWERLEVEL9K_LEFT_PROMPT_ELEMENTS=(context dir virtualenv vcs)
POWERLEVEL9K_RIGHT_PROMPT_ELEMENTS=(status history time)
POWERLEVEL9K_STATUS_VERBOSE=false
POWERLEVEL9K_LEFT_SEGMENT_SEPARATOR=''
POWERLEVEL9K_RIGHT_SEGMENT_SEPARATOR=''
POWERLEVEL9K_PROMPT_ON_NEWLINE=true
POWERLEVEL9K_MULTILINE_FIRST_PROMPT_PREFIX=''
POWERLEVEL9K_MULTILINE_SECOND_PROMPT_PREFIX='$ '

export EDITOR=`which vim`

export VAGRANT_DEFAULT_PROVIDER='vmware_fusion'

# history substring search: bind UP and DOWN arrow keys
zmodload zsh/terminfo
bindkey "$terminfo[kcuu1]" history-substring-search-up
bindkey "$terminfo[kcud1]" history-substring-search-down

# zsh syntax highlighting
ZSH_HIGHLIGHT_HIGHLIGHTERS=(main brackets pattern cursor)

source /usr/local/bin/virtualenvwrapper.sh
. `brew --prefix`/etc/profile.d/z.sh

alias scpresume="rsync --partial --progress --rsh=ssh"
alias ssh="eval $(keychain --eval --agents ssh -Q --quiet id_rsa) ssh"
```

**Comment by dritter** [comment]
I can reproduce this bug even in `master`.. If I change my theme to any other, it disappears. So it must be a problem of powerlevel9k. :(

**Comment by jmiguelv** [comment]
Just rolled back to version 0.1.0 and the history substring search plugin works with that version.

**Comment by jmiguelv** [comment]
It seems the plugin stopped working after commit f18c92d76664f53351515f0f71a1047b7766648c

**Comment by jmiguelv** [comment]
That commit added 2 new functions `zle-line-init` and `zle-keyma-select`, they both appear to do the same, preparing and resetting the prompts. Commenting out `zle-line-init` seems to fix the issue, but I am not sure why.

**Comment by dritter** [comment]
@jmiguelv Great work! Thanks. I didn't expect that..

**Comment by fbbdev** [comment]
`history-substring-search` (and many other plugins) depend on oh-my-zsh changing terminal mode in `zsh-line-init`. See robbyrussell/oh-my-zsh#4540
Maybe you can check if the function is already defined and wrap it in your implementation.

**Comment by bhilburn** [comment]
@dritter - Are you following up on this one (I just don't want to duplicate work)? I'd like to get a fix merged into `master` as soon as we can.

**Comment by dritter** [comment]
Well.. I tried to adapt the fixes in the according pull requests of OMZ but had no luck. I am currently waiting for them to see how they fix this problem. I am not much into keymaps, so I am not the best one on that.

**Comment by bhilburn** [comment]
I'm actually fairly confused about this, because `history-substring-search` still works for me using powerlevel9k `next` branch, and latest ZSH pull.

@jmiguelv - What version of ZSH are you using? I'm on e44aa50301

**Comment by jmiguelv** [comment]
@bhilburn I am on `zsh 5.1.1 (x86_64-apple-darwin15.0.0)`

**Comment by bhilburn** [comment]
@jmiguelv - Judging by the fact that it doesn't work for you or @dritter, and you are both on Macs, I'm guessing that's the difference - I'm on Linux.

@dritter - This works perfectly fine for me. Any guesses on what might be different between the OSes that could be causing this?

I'm on `zsh 5.0.8 (x86_64-redhat-linux-gnu)`. I wonder if the older version of ZSH is impacting this.

**Comment by dritter** [comment]
@bhilburn I still can't believe this is a OS thing. Rather a configuration issue. I can see you set the `history-substring-search-up` and `history-substring-search-down` keys explicitly. Maybe this is why it works for you? But this is just a wild guess..

**Comment by bhilburn** [comment]
@dritter - If I remove those keybindings, it does indeed stop working.

That raises the question, though, of how our theme could possibly break that?

**Comment by jmiguelv** [comment]
What is the main reason for the theme to implement `zle-line-init`? It
seems to work fine without it, that's how I have been using it to
workaround this issue.

On 4 Nov 2015 19:12, "Ben Hilburn" <notifications@github.com> wrote:
>
> @dritter - If I remove those keybindings, it does indeed stop working.
>
> That raises the question, though, of how our theme could possibly break
that?

**Comment by dritter** [comment]
@jmiguelv The main reason is the `vi_mode` segment. Without that function the initial state of the vi-mode won't be displayed. Also, if you press CTRL+C the vi-mode segment disappears..

**Comment by fbbdev** [comment]
I don't know zsh very much, but can't you do something like this
(pseudo-code):

‘‘‘
if zle-line-init is defined
  zle-line-init-old = body of zle-line-init
end if

zle-line-init () {
  zle-line-init-old
  # powerlevel9k settings
  ...
}
‘‘‘

**Comment by bhilburn** [comment]
I'd like to get a fix merged into master for this as soon as we can - I imagine this could annoy folks quite a bit.

@dritter - Any thoughts on @fbbdev's suggestion? Have any other themes solved this issue, I wonder?

**Comment by dritter** [comment]
I really don't think this is an issue of this theme. Rather an issue of the ZSH settings the user has set (or of the framework the user uses).

In the [zsh-history-substring-search](https://github.com/zsh-users/zsh-history-substring-search) plugins readme it says to set:

```zsh
# bind UP and DOWN arrow keys
zmodload zsh/terminfo
bindkey "$terminfo[kcuu1]" history-substring-search-up
bindkey "$terminfo[kcud1]" history-substring-search-down

# bind UP and DOWN arrow keys (compatibility fallback
# for Ubuntu 12.04, Fedora 21, and MacOSX 10.9 users)
bindkey '^[[A' history-substring-search-up
bindkey '^[[B' history-substring-search-down

# bind P and N for EMACS mode
bindkey -M emacs '^P' history-substring-search-up
bindkey -M emacs '^N' history-substring-search-down

# bind k and j for VI mode
bindkey -M vicmd 'k' history-substring-search-up
bindkey -M vicmd 'j' history-substring-search-down
```
That did the trick for me. I don't understand enough of the different keymappings ZSH has to come up with a serious solution, but I think this is a workaround we don't want to set in the theme.

@fbbdev I am not sure if your solution would work, but IMHO it would feel strange to have it like that.

**Comment by jmiguelv** [comment]
I already had the following in my `.zshrc`:
```
# bind UP and DOWN arrow keys
zmodload zsh/terminfo
bindkey "$terminfo[kcuu1]" history-substring-search-up
bindkey "$terminfo[kcud1]" history-substring-search-down
```
which worked with other themes. If I change the configuration to
```
# bind UP and DOWN arrow keys (compatibility fallback
# for Ubuntu 12.04, Fedora 21, and MacOSX 10.9 users)
bindkey '^[[A' history-substring-search-up
bindkey '^[[B' history-substring-search-down
```
it also works with powerlevel9k. Thank you for the hint @dritter.

**Comment by dritter** [comment]
@apjanke gives an explanation in https://github.com/robbyrussell/oh-my-zsh/issues/2735

**Comment by apjanke** [comment]
Yep, that's what it looks like: redefining `zle-line-init` and not having it issue `smkx` means  `$terminfo`-based key mappings for keypad buttons can't be used. `terminfo` only supplies the "keypad application" mode sequences. The hardcoded `'^[[A'` and `^[[B'` sequences are what xterms send in "keypad local" mode.

I think you'd want to support `$terminfo`. It's nice for portability and readability, and it seems to be what various Zsh users are converging on.

@fbbdev's approach of wrapping the existing function feels hacky, but it should work. You can get the old function body with `which zle-line-init`, replace the function name at the beginning, and `eval` it to define the `zle-line-init-orig` function. Probably need to do the same for `zle-keymap-select`. That seems a little gnarly, though. It'd be nice if there were an equivalent of `$precmd_functions` for multiple `zle-line-init` hooks to be registered.

**Comment by dritter** [comment]
@apjanke I have some trouble getting that to work. I see in your PR https://github.com/robbyrussell/oh-my-zsh/pull/4191 that you removed the `smkx` and `rmkx` related code..

My simple approach was:
```zsh
function zle-line-init {
   powerlevel9k_prepare_prompts
   if (( ${+terminfo[smkx]} )); then
     printf '%s' ${terminfo[smkx]}
   fi
   zle reset-prompt
   zle -R
}
```
But that didn't work. 
Thanks for your help.

**Comment by apjanke** [comment]
@dritter The code I removed in robbyrussell/oh-my-zsh#4191 was some broken additional `smkx`/`rmkx` code that was overriding the [original `zle-line-init` in the core OMZ `lib/key-bindings.zsh` files](https://github.com/robbyrussell/oh-my-zsh/blob/83c4e0620689d2410056bddcdfbf8e300673e982/lib/key-bindings.zsh#L5-L16) (Notice how it unconditionally emitted `rmkx` immediately after `smkx`?) The current state of the `vi-mode` plugin there means that loading it causes OMZ users to encounter similar issues to what folks are seeing here in this issue.

Your example code with `smkx` seems to work for me. (Though it should probably include a corresponding `zle-line-finish` which emits `rmkx`.) I downloaded powerlevel9k and changed my `~/.zshrc` to this simple one.

```
source ~/local/opp/powerlevel9k/powerlevel9k/powerlevel9k.zsh-theme

# bind UP and DOWN arrow keys
zmodload zsh/terminfo
bindkey "$terminfo[kcuu1]" history-substring-search-up
bindkey "$terminfo[kcud1]" history-substring-search-down
```

With the current version of Powerlevel9k, pressing up arrow while editing the command line emits `^[[A` and those bindings using terminfo do not work. (You can test it by doing Ctrl-V and then pressing an arrow key.) With your changed `zle-line-init` that does `smkx`, up arrow emits `^[0A`, and those bindings work. (Causing the arrows to complain `No such widget `history-substring-search-up'`, because I don't have history-substring-search installed. But that means the key bindings are working as intended.)

(This is on OS X using Terminal.app or iTerm2.)

**Comment by dritter** [comment]
@apjanke Perfect! That works for me (at least in a plain ZSH). The reason why that didn't work in my tests before was, that I used OMZ (which has a broken config that you try to fix in your PRs). Thanks a lot!

**Comment by apjanke** [comment]
Excellent! Glad to be of help.

**Comment by bhilburn** [comment]
Fixed by PR #136.

Great work, @dritter!!
